### DISCUSSION TITLE
Testing with useEffect - only called once

### DISCUSSION THREAD

**Discussion by mcarter1** [react-hook-form/react-hook-form#6876] [Q&A]
Hi all,

I'm new to React testing library and react-hook-form, so hopefully there's a simple answer to this. 

I have a simple search page which is calling useEffect. This can set an error message to be displayed at the top of the form, depending on the values set in the search fields.

```
  useEffect(() => {
    if (...) {
            setState({ 
              error: (getText('surname-search')), 
              result: null })
    }
  }
```
All works fine in the UI, but now I need to write a test for it.
```
    const searchBtn = component.getByTestId('search-button')
    fireEvent.submit(searchBtn);
    expect(handleSubmit).toHaveBeenCalled;
    const validation = component.getByText("error text");
    expect(validation).toBeInTheDocument;
```
The error message I'm looking for isn't found, and doesn't appear on the logged DOM. After a bit of debugging, I find that running the test, the useEffect is called and runs OK on render, but doesn't get called again after I submit through the search button. I do see handleSubmit being called.

I don't really want to mock the useEffect as it contains logic for determining when to set an error message and what the message should be.

Any ideas?

**Comment by JamesWong1999** [comment]
are you wrapping your change with `act`?

```
await act(async () => {
  fireEvent...
})
```

**Reply by mcarter1** [reply]
I've tried it either way, I didn't think act was required for React testing library.

This isn't an async component.

**Comment by j0hannr** [comment]
I am not sure what you want to achieve...

### Suggestion A
Use the custom Validation function found [here in the docs](https://react-hook-form.com/get-started#Handleerrors)

### Suggestion B
Run your testing script 

```js
const onSubmit = async (form) => {
   // your script
}
```
